src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1381,6 +1382,94 @@
   (let [last-modified-milliseconds (.lastModified (io/file cacrl))]
        (time-coerce/from-long last-modified-milliseconds)))
 
+(schema/defn ^:always-validate reject-delta-crl
+  [crl :- CertificateRevocationList]
+  (when (utils/delta-crl? crl)
+    (throw (IllegalArgumentException.
+            ^String (i18n/trs "Cannot support delta CRL.")))))
+
+(schema/defn ^:always-validate validate-certs-and-crls
+  "Given a list of certificates and a list of CRLs, validate the certificate
+   chain, i.e. ensure that none of the certs have been revoked by checking the
+   appropriate CRL, which must be present and currently valid. Delta CRLs are
+   not supported. Returns nil if successful."
+  [cert-chain :- [Certificate]
+   crl-chain :- [CertificateRevocationList]]
+  (doseq [crl crl-chain]
+    (reject-delta-crl crl))
+  (try
+    (utils/validate-cert-chain cert-chain crl-chain)
+    ;; currently not distinguishing between invalid CRLs
+    ;; and a revoked cert in the CA chain
+    (catch CertPathValidatorException e
+      (throw (IllegalArgumentException.
+              ^String (i18n/trs "Invalid certs and/or CRLs: {0}" (.getMessage e)))))))
+
+(schema/defn ^:always-validate get-newest-crl :- CertificateRevocationList
+  "Determine the newest CRL by looking for the highest CRL Number. This assumes
+  all given CRLs have the same issuer. Fails if more than one CRL has the
+  highest CRL Number."
+  [crls :- [CertificateRevocationList]]
+  (let [newest-crl (->> crls
+                        (group-by utils/get-crl-number)
+                        (into (sorted-map))
+                        last
+                        last)]
+    (if (= 1 (count newest-crl))
+      (first newest-crl)
+      (throw (IllegalArgumentException.
+              ^String (i18n/trs "Could not determine newest CRL."))))))
+
+(schema/defn ^:always-validate maybe-replace-crl :- CertificateRevocationList
+  "Given a CRL and a map of key identifiers to CRLs, determine the
+  newest CRL with the key-id of the given CRL. Warn if the newest CRL
+  is the given CRL."
+  [crl :- CertificateRevocationList
+   key-crl-map :- {schema/Any [CertificateRevocationList]}
+   ca-name :- schema/Str]
+  (if (= ca-name (utils/get-cn-from-x500-principal
+                  (.getIssuerX500Principal crl)))
```

Comment:
Should we be checking the keyi d here?

Reply:
Yes, probably. Need to figure out a good way to find the right cert to extract it from.

Reply:
Still need to do this.